### PR TITLE
Add file edit options on frontend

### PR DESCRIPTION
**Pull request by nickbar01234**
# Description

- Add options to edit files on frontend
- Allow chapter leaders to edit files if they're assigned to a senior
- Update CRUD API for files

### CODE REVIEW

File: prisma/schema.prisma [hunk 1]
```diff
@@ -105,7 +105,11 @@ model File {
   id       String   @id @default(auto()) @map("_id") @db.ObjectId
   date     DateTime // will zero out the hours
   filetype String
+
+  // Use with Google API
+  fileId   String
```
Comment (nickbar01234) [inline prisma/schema.prisma:110]: Make it easier to keep track instead of parsing the string.

File: src/app/api/senior/[id]/route.ts [hunk 2]
```diff
@@ -106,13 +106,14 @@ export const PATCH = withSessionAndRole(
       );
       const studentIDsToAdd = seniorBody.StudentIDs;
 
+      const { StudentIDs: _, ...other } = seniorBody;
```
Comment (nickbar01234) [inline src/app/api/senior/[id]/route.ts:109]: This actually was a bug we didn't notice before. Without destructuring `StudentIDs`, the relation on `User` side is not updated.

File: src/app/api/file/[fileId]/route.ts [hunk 3]
```diff
@@ -26,63 +25,32 @@ export const PATCH = withSession(async (request) => {
...
-    // Check that user has this senior assigned to them
-    const user = await prisma.user.findFirst({
+    const maybeFile = await prisma.file.findFirst({
       where: {
-        id: request.session.user.id,
+        id: fileId,
+      },
+      include: { senior: true },
+    });
+    const otherFiles = await prisma.file.findMany({
```
Comment (nickbar01234) [inline src/app/api/file/[fileId]/route.ts:34]: Without this check, user can make files with the same date!

File: src/components/file/AddFile.tsx [hunk 4]
```diff
@@ -6,22 +6,33 @@ import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
...
   seniorFolder,
   files,
+  editFile,
+  setEditFile,
 }: {
   seniorId: string;
   seniorFolder: string;
   files: PrismaFile[];
+  editFile?: PrismaFile;
+  setEditFile: React.Dispatch<React.SetStateAction<PrismaFile | undefined>>;
```
Comment (nickbar01234) [inline src/components/file/AddFile.tsx:22]: I don't think this is too clean, but it makes the most sense given the time constraint.

File: src/components/user/AddFile.tsx [hunk 5]
```diff
@@ -168,19 +200,28 @@ const AddFile = ({
...
-                body: {
-                  date: startDate,
-                  filetype: "Google Document",
-                  url: "",
-                  Tags: selectedTags.map((tagProp) => tagProp.name),
-                  seniorId: seniorId,
-                },
-              });
+            onClick={async () => {
+              editFile == null
```
Comment (wkim10) [inline src/components/user/AddFile.tsx:204]: ===

File: src/components/user/AddFile.tsx [hunk 6]
```diff
@@ -88,35 +91,64 @@ const AddFile = ({
...
+        } else {
+          setError(true);
+        }
+      },
+    });
+
+  const fetching = fetchingCreateFile || fetchingUpdateFile;
+
+  React.useEffect(() => {
+    if (editFile != undefined) {
```
Comment (wkim10) [inline src/components/user/AddFile.tsx:144]: !==

File: src/components/user/AddFile.tsx [hunk 7]
```diff
@@ -168,19 +200,28 @@ const AddFile = ({
               selectedTags.length == 0 ||
```
Comment (wkim10) [inline src/components/user/AddFile.tsx:200]: this wasn't one of your changes but triple equals ===

File: src/components/file/File.tsx [hunk 8]
```diff
@@ -1,21 +1,78 @@
...
+  const router = useRouter();
+
+  const { fetching: fetchingDeleteFile, fn: throttleDeleteFile } =
+    useApiThrottle({
+      fn: deleteFile,
+      callback: () => router.refresh(),
+    });
+
+  const fetching = fetchingDeleteFile;
+  const editable = setFileEdit != undefined;
```
Comment (wkim10) [inline src/components/file/File.tsx:36]: !==

File: src/components/file/AddFile.tsx [hunk 9]
```diff
@@ -6,22 +6,33 @@ import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
...
 }) => {
   const [showAddFilePopUp, setShowAddFilePopUp] = useState<boolean>(false);
 
   const handlePopUp = () => {
     setShowAddFilePopUp(!showAddFilePopUp);
   };
 
+  React.useEffect(() => {
+    if (editFile != null) {
```
Comment (wkim10) [inline src/components/file/AddFile.tsx:31]: !==

File: src/app/api/file/[fileId]/route.ts [hunk 10]
```diff
@@ -128,68 +79,32 @@ export const PATCH = withSession(async (request) => {
...
-      url: `https://docs.google.com/document/d/${fileId}`,
+      id: fileId,
+    },
+    include: {
+      senior: true,
     },
   });
 
-  if (file === null) {
+  if (file == null || !file.senior.StudentIDs.includes(session.user.id)) {
```
Comment (wkim10) [inline src/app/api/file/[fileId]/route.ts:95]: ===

**Review by johnny-t06 — CHANGES_REQUESTED** [review]
Nice work Nick! Changes requested for the error codes and maybe a later ticket for user feed back

**Review by johnny-t06 — APPROVED** [review]
Query logic works great. Great changes, should be good to merge!

**Comment by nickbar01234** [comment]
Resolved with @wkim10 on `null` vs `undefined` equality check. I think we're good for now!

**Pull request merged**